File: app/src/main/java/com/example/influencer/Domain/LoginUseCase.java
package com.example.influencer.Domain;

import androidx.annotation.NonNull;

import com.example.influencer.Data.Network.AuthenticationService;
import com.example.influencer.UI.Login.LoginListener;
import com.example.influencer.UI.Login.Model.UsuarioLogin;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;

import javax.inject.Inject;

//aca lo 100% ideal seria o ya sea usar RxJava (un Single (equivalante al floawble pero para 1 unica operacion asyncronica nada mas), Completable, maybe) o corrutinas de Kotlin
//ya que al usar callBacks si bien el UseCase no maneja la implementacion de "task.isSuccessful()" si es conciente de la existencia de la interfaz (Listener), que seria un componente de
//que lo implementa el viewmodel, aun asi, dentro de todo sigue cumpliendo un MVVM con clean Architecure ya que depsues de todo no sabe ninguna implementacion (todo lo hace el viewmodel)
public class LoginUseCase {
    private final AuthenticationService authenticationService;

    @Inject
    public LoginUseCase(AuthenticationService authenticationService) {
        this.authenticationService = authenticationService;
    }

    public void invoke(UsuarioLogin usuarioLogin, LoginListener listener) {
        authenticationService.login(usuarioLogin.getEmail(), usuarioLogin.getContrasena()).addOnCompleteListener(new OnCompleteListener<AuthResult>() {
            @Override
            public void onComplete(@NonNull Task<AuthResult> task) {
                if (task.isSuccessful()){
                    listener.LoginSuccess();
                }else{
                    listener.LoginError();
                }
            }
        });
    }
}

File: app/src/main/java/com/example/influencer/Features/Login/Domain/LoginUseCase.kt
package com.example.influencer.Features.Login.Domain

import javax.inject.Inject
import com.example.influencer.Core.Data.Network.AuthenticationService
import com.example.influencer.Features.Login.Domain.Model.UsuarioLogin
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

//aca lo 100% ideal seria o ya sea usar RxJava (un Single (equivalante al floawble pero para 1 unica operacion asyncronica nada mas), Completable, maybe) o corrutinas de Kotlin
//ya que al usar callBacks si bien el UseCase no maneja la implementacion de "task.isSuccessful()" si es conciente de la existencia de la interfaz (Listener), que seria un componente de
//que lo implementa el viewmodel, aun asi, dentro de todo sigue cumpliendo un MVVM con clean Architecure ya que depsues de todo no sabe ninguna implementacion (todo lo hace el viewmodel)
class LoginUseCase @Inject constructor(
    private val authenticationService: AuthenticationService
) {
    suspend operator fun invoke(usuarioLogin: UsuarioLogin) {
        return suspendCancellableCoroutine { continuation ->
            authenticationService.login(usuarioLogin.email, usuarioLogin.contrasena)
                .addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        continuation.resume(Unit)
                    } else {
                        continuation.resumeWithException(task.exception ?: Exception("Login failed"))
                    }
                }
        }
    }
}